runtime/src/iree/hal/drivers/metal/metal_device.m

```diff
@@ -446,9 +499,13 @@ static iree_status_t iree_hal_metal_device_queue_execute(
                                            value:signal_semaphore_list.payload_values[i]];
       }
 
+      // Retain the block pool to make sure it outlive the resource set we created in the above.
+      iree_hal_metal_arena_block_pool_t* pool = device->block_pool;
```

Comment:
behavior is undefined in the HAL if any device resources lives longer than the device itself - this looks like a workaround for that happening, which should be fixed instead by keeping the device alive (which owns the block pool) if you can't otherwise ensure that a device is not destroyed with pending callbacks (by calling a MTLDevice wait-until-all-callbacks-issued or whatever)

Reply:
Yeah I was thinking about doing it that way too. Let me try to see if there is a clean way.

Reply:
Another attempt: https://github.com/openxla/iree/pull/15288 :)